### ISSUE TITLE
Add `--forward` flag to `etl` to control dependencies

### ISSUE THREAD

**Issue by Marigold** [owid/etl#239]
## Problem

Running `etl data://garden/wb/2021-07-01/wb_income --dry-run` would currently rebuild all _backward_ and _forward_ steps. Here's a complete list

```
Detecting which steps need rebuilding...
Running 16 steps:
1. walden://faostat/2021-03-18/faostat_QCL...
2. walden://faostat/2022-02-10/faostat_metadata...
3. walden://faostat/2017-12-11/faostat_FBSH...
4. walden://faostat/2021-04-09/faostat_FBS...
5. data://open_numbers/open_numbers/latest/open_numbers__world_development_indicators...
6. data://meadow/faostat/2021-03-18/faostat_qcl...
7. data://meadow/faostat/2022-02-10/faostat_metadata...
8. data://meadow/faostat/2017-12-11/faostat_fbsh...
9. data://meadow/faostat/2021-04-09/faostat_fbs...
10. data://garden/owid/latest/key_indicators...
11. data://garden/faostat/2021-03-18/faostat_qcl...
12. data://garden/faostat/2021-04-09/faostat_fbsc...
13. data://open_numbers/open_numbers/latest/gapminder__systema_globalis...
14. data://garden/owid/latest/population_density...
15. data://garden/explorers/2021/food_explorer...
16. data://examples/script/script_example...
```

By backward dependencies we mean all steps that we depend on, i.e.
```
data://meadow/wb/2021-07-01/wb_income
walden://wb/2021-07-01/wb_income
```
and by forward dependencies all steps that depend on `data://garden/wb/2021-07-01/wb_income`. That is `data://garden/owid/latest/key_indicators` and then forward deps on `key_indicators`, etc.

Once we get all backward and forward dependencies, we then also include [all their backward dependencies](https://github.com/owid/etl/blob/112e5c4c27203c34df7c2ed136e003983d7aa6e1/etl/steps/__init__.py#L84-L86) to make sure our subgraph is "complete" (might not be an official graph theory term). This is how we get all those 16 steps above.

This is useful if you want to rebuild a specific step and make sure everything that would change gets updated, but we rarely do that since in production we always run `etl` and execute full dag. On the other hand, we prototype a lot locally and frequently need to run specific steps and their backward dependencies. Running forward dependencies just slows it down.

## Solution

We should always run backward dependencies automatically, but we could have `--forward` flag (off by default) to control what gets build.

An alternative is having `--only` that would run only steps matching `etl` regex. Another is implementing both `--backward` and `--forward` (then `--only` is the same as running it without `--backward` and `--forward`), but I can't think of any scenario where excluding `--backward` would be helpful to my dev workflow.

Currently one could also use `--exclude` flag, but that becomes impractical with complex dag (like the one above).

**Comment by Marigold** [comment]
cc @larsyencken @pabloarosado

**Comment by wcox13** [comment]
It seems best to me to add the simplest flag that addresses the need, for now anyway. So I'd vote the `--forward` option.

That would basically just be a matter of enclosing [this logic](https://github.com/owid/etl/blob/112e5c4c27203c34df7c2ed136e003983d7aa6e1/etl/steps/__init__.py#L84-L86) in an `if` statement controlled by the `--forward` flag, right?

**Comment by wcox13** [comment]
Also, if we had `--forward` off by default, would there be any problem with existing behavior? Should we have it on by default and add `--no-forward` to preserve current defaults?

**Comment by Marigold** [comment]
> That would basically just be a matter of enclosing [this logic](https://github.com/owid/etl/blob/112e5c4c27203c34df7c2ed136e003983d7aa6e1/etl/steps/__init__.py#L84-L86) in an if statement controlled by the --forward flag, right?

I think you'll rather have to make changes to [filter_to_subgraph](https://github.com/owid/etl/blob/112e5c4c27203c34df7c2ed136e003983d7aa6e1/etl/steps/__init__.py#L82). That logic you mention is taking care of backward dependencies. You might even have to restructure `filter_to_subgraph` a bit.

> Also, if we had --forward off by default, would there be any problem with existing behavior?

Don't think so. In production we use `etl` which computes all steps and in staging we use `etl garden explorer`, but there are no other forward steps depending on those. (And if there ever are we would explicitly enumerate them)

**Comment by wcox13** [comment]
Refactor PR, which I think simplifies a good bit: https://github.com/owid/etl/pull/244

After that, it's trivial to add the command line arg / change the default, just saving the behavior change for a separate PR

**Comment by Marigold** [comment]
So... beautiful... diagrams. How did you make it @larsyencken? Tell me!

> consistent subgraph

Just out of curiosity, is this an official definition? I was using _maximal subgraph_, but that was totally made up...

Anyway, I agree with everything...

**Comment by larsyencken** [comment]
Haha, I've been playing with Excalidraw: https://excalidraw.com/#json=xsQWYQbVXJrL2v8nAA2is,xeP3Je0Era_-yDVYaaEztA.

I've been looking at graph theory terms, I don't think there's a neat one for what we're describing, which is specific to DAGs.

https://en.wikipedia.org/wiki/Glossary_of_graph_theory#:~:text=A%20connected%20component%20of%20a,components%2C%20and%20strongly%20connected%20components.

A _closure_ is a subgraph with no edges going outside the graph. A _connected component_ is a subgraph where there is a path from every node to another, and no paths leading outside the component. But... it's hard to describe what we're doing exactly in those terms 😅

**Comment by larsyencken** [comment]
Ah, yeah, I'm not sure "dependents" is very natural either, unfortunately. Trying on a few:

- `--forward`
- `--downstream`
- `--descendants`
- `--complete`

I'm not so sure what's easiest to understand.

**Comment by wcox13** [comment]
New flag + new default behavior: https://github.com/owid/etl/pull/248

**Comment by pabloarosado** [comment]
Sorry for joining late. Some comments:
* I personally find "backward" and "forward" the most intuitive words, because it's clear (in time) that backward steps are ones that need to happen before, and forward steps are those that need to happen later. Downstream is a little confusing because, firstly, it's not clear in time what is above or below, and secondly, in the dag, we put the **upstream** dependencies **below** the step. But if you have already agreed on this nomenclature, I'm fine with it.
* I like Lars' diagrams, very clear. I think the future behaviour is indeed the most meaningful default behaviour. We should keep in mind that, while developing a data pipeline, it's useful to have the option to run the minimum amount of code possible, since you are constantly changing things and don't want to waste time. So I think the "only" option (to execute just the selected steps) is definitely useful.

Thanks for doing this, it will be very useful!

**Comment by Marigold** [comment]
> So I think the "only" option (to execute just the selected steps) is definitely useful.

I wouldn't implement it yet and wait how avoiding forward steps works for us. In my opinion it should be enough for most (maybe even all) practical purposes.

**Comment by pabloarosado** [comment]
The "only" option is useful while developing a pipeline, because you may be simultaneously working on different steps (some of them may depend on others, and they may take some time to execute). You already have existing datasets for all of them (they are not up-to-date, but they exist), and you want to check that a specific feature of a specific step works. You don't care if the data that this step is ingesting is the latest, you just want to run it with any data to see that it works.

**Comment by Marigold** [comment]
@pabloarosado thanks, this is indeed a good use case. @wcox13 would you mind either adding `--only` option to this PR or open a new one if the change is more complex?

**Comment by wcox13** [comment]
Thanks Pablo, yup I'll open a new PR for `--only`!

**Comment by wcox13** [comment]
Actually, I just added to the same PR, ready for review: https://github.com/owid/etl/pull/248

**Comment by wcox13** [comment]
This is done actually

**Comment by wcox13** [comment]
Can’t figure out how to close it though, maybe don’t have permission?
